### PR TITLE
Change exoplayer payload type

### PR DESCRIPTION
**Pull request by uvjustin**
## Proposed change

The frontend may need to send additional options to the exoplayer/play_hls interface such as whether to be muted or not: home-assistant/frontend#6857 . With that in mind it is probably better to use a JSONObject instead of a String for the payload.
Have also opened up a corresponding PR on the Android side (https://github.com/home-assistant/android/pull/932). If we make the change it is probably better to do it before 0.115 before any of the exoplayer stuff is officially added.

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (thank you!)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

```

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #6857
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by bramkragten** [comment]
This will be breaking, how do we handle an old version of the Android app?

**Comment by uvjustin** [comment]
We could change the External Config tag from something like hasExoPlayer to hasExoPlayerv2.
But this might be unnecessary clutter - if Android side gets pulled and the app gets updated before the frontend change there will be minimal disruption. The ExoPlayer code has only been in Android for ~ 2 weeks and the frontend side hasn't had it in a full release yet. It will be more of a headache if we make the change after 0.115 gets released.

**Comment by bramkragten** [comment]
The Android app is out already, how can we be sure users update it before Thursday? That's not much time...

**Comment by uvjustin** [comment]
That's a good point. If the app doesn't get updated in time then we should probably just skip exoplayer for 0.115 and wait until the next version. I don't think it's worth cluttering up anything for a workaround.

**Comment by bramkragten** [comment]
Ok, let's do that then

**Comment by uvjustin** [comment]
I'll make a small PR to return false for the ExoPlayer check. When the Android App gets updated we can reenable it, even in a minor version.

**Comment by uvjustin** [comment]
@bramkragten The Android app was updated in the store a few days ago, so this can be merged without breaking now.

File: src/components/ha-hls-player.ts [hunk 1]
```diff
@@ -34,8 +35,7 @@ class HaHLSPlayer extends LitElement {
   @property({ type: Boolean, attribute: "playsinline" })
   public playsInline = false;
 
-  @property({ type: Boolean, attribute: "allow-exoplayer" })
-  public allowExoPlayer = false;
```
Comment (bramkragten) [inline src/components/ha-hls-player.ts:38]: Why this change? You can't use an attribute now to set the property?

Comment (uvjustin) [reply src/components/ha-hls-player.ts:38]: It seemed like the attribute was not necessary and it works fine without it. As I commented before I'm not sure it was even working before as there is no attribute `allow-exoplayer` coming down from `ha-camera-stream` (`more-info-camera` passes down `allow-exoplayer` to `ha-camera-stream` as an attribute, but `ha-camera-stream` passes down `allowExoPlayer` to `ha-hls-player` as a property, so looking for the `allow-exoplayer` attribute in `ha-hls-player` comes up false. I'm a JS novice so maybe I'm missing something with the property to attribute relationship, but if it works this way isn't it more parsimonious?

Comment (bramkragten) [reply src/components/ha-hls-player.ts:38]: No, it limits the options and people expect it to be available as an attribute.

It will just work if you pass it as a property instead of an attribute.

Comment (uvjustin) [reply src/components/ha-hls-player.ts:38]: I see, so better to have it available as an attribute in case someone wants to access the `ha-hls-player `directly and not through `ha-camera-stream`?

**Comment by uvjustin** [comment]
In 115.5 we made a change to stream to specify the codecs in a HLS master playlist. Since we now have the codecs available in the master playlist, we are now able to check the master playlist for H.265 before resorting to Exoplayer. I've just pushed the relevant changes.
Note we can also do this through the HLS library but it will be a little more complex, plus there's a change that needs to be pushed to the DefinitelyTyped library to fix one of the event types.

File: src/components/ha-hls-player.ts [hunk 2]
```diff
@@ -93,34 +94,40 @@ class HaHLSPlayer extends LitElement {
...
-      if (!hlsSupported) {
-        this._videoEl.innerHTML = this.hass.localize(
-          "ui.components.media-browser.video_not_supported"
-        );
-        return;
-      }
+    const hls = ((await import(
+      /* webpackChunkName: "hls.js" */ "hls.js"
+    )) as any).default as HLSModule;
+    let hlsSupported = hls.isSupported();
```
Comment (uvjustin) [inline src/components/ha-hls-player.ts:113]: Note we aren't modifying hlsSupported with this._useExoPlayer. It's probably redundant. Support for ExoPlayer should pretty much be a subset of hls.isSupported().

File: src/components/ha-hls-player.ts [hunk 3]
```diff
@@ -93,34 +94,40 @@ class HaHLSPlayer extends LitElement {
...
+        "ui.components.media-browser.video_not_supported"
+      );
+      return;
+    }
 
-    if (this._useExoPlayer) {
+    this._useExoPlayer = await useExoPlayerPromise;
+    const hevcRegexp = /CODECS=".*?((hev1)|(hvc1))\..*?"/;
+    const masterPlaylist = await masterPlaylistPromise.text();
+    if (hevcRegexp.test(masterPlaylist) && this._useExoPlayer) {
```
Comment (bramkragten) [inline src/components/ha-hls-player.ts:130]: Shouldnt we do this before loading and checking HLS?

Comment (uvjustin) [reply src/components/ha-hls-player.ts:130]: We could, and doing so would avoid the unnecessary loading of the HLS module in the case that we end up using ExoPlayer. But moving it up to before the HLS module load would likely be slower for the average case (I assume most cases will still use the hls.js polyfill or native) since we would block while waiting for the url to be fetched. Let me know which way you prefer.

File: src/components/ha-hls-player.ts [hunk 4]
```diff
@@ -93,34 +94,40 @@ class HaHLSPlayer extends LitElement {
...
+      this._videoEl.innerHTML = this.hass.localize(
+        "ui.components.media-browser.video_not_supported"
+      );
+      return;
+    }
 
-    if (this._useExoPlayer) {
+    this._useExoPlayer = await useExoPlayerPromise;
+    const hevcRegexp = /CODECS=".*?((hev1)|(hvc1))\..*?"/;
+    const masterPlaylist = await masterPlaylistPromise.text();
```
Comment (bramkragten) [inline src/components/ha-hls-player.ts:129]: We should skip these 2 if `this._useExoPlayer === false`?

File: src/components/ha-hls-player.ts (hunk 3 above)
Comment (bramkragten) [reply src/components/ha-hls-player.ts:130]: OK fair enough :-)

File: src/components/ha-hls-player.ts (hunk 4 above)
Comment (uvjustin) [reply src/components/ha-hls-player.ts:129]: Yup, let me change it

File: src/components/ha-hls-player.ts (hunk 3 above)
Comment (uvjustin) [reply src/components/ha-hls-player.ts:130]: Whoops, think it is blocking as is. The fetch() function involves network while the text() function should be fast right? Let me move the first await down.

File: src/components/ha-hls-player.ts [hunk 5]
```diff
@@ -93,35 +94,43 @@ class HaHLSPlayer extends LitElement {
...
+    }
 
+    this._useExoPlayer = await useExoPlayerPromise;
     if (this._useExoPlayer) {
-      this._renderHLSExoPlayer(url);
+      const hevcRegexp = /CODECS=".*?((hev1)|(hvc1))\..*?"/;
+      const masterPlaylist = await (await masterPlaylistPromise).text();
+      if (hevcRegexp.test(masterPlaylist)) {
+        this._renderHLSExoPlayer(url);
+      }
```
Comment (bramkragten) [inline src/components/ha-hls-player.ts:133]: We will not fallback to the default hls player now

**Comment by uvjustin** [comment]
The changes we made in 0.115.5 also changed the url from a regular playlist to a master playlist. The various players don't need the master playlist to play - we only added it because Google Cast devices needed the codec strings to function properly. Should we just send the regular playlist instead?

**Comment by bramkragten** [comment]
> The changes we made in 0.115.5 also changed the url from a regular playlist to a master playlist. The various players don't need the master playlist to play - we only added it because Google Cast devices needed the codec strings to function properly. Should we just send the regular playlist instead?

Uuuuh... I would not know? @hunterjm?

**Comment by uvjustin** [comment]
It should be a little faster if we do that because we skip one extra master playlist load. The only concern is that we are coupling tight to the backend and have to remember if we ever change anything there.

File: src/components/ha-hls-player.ts [hunk 6]
```diff
@@ -125,12 +125,15 @@ class HaHLSPlayer extends LitElement {
...
-      const hevcRegexp = /CODECS=".*?((hev1)|(hvc1))\..*?"/;
-      const masterPlaylist = await (await masterPlaylistPromise).text();
-      if (hevcRegexp.test(masterPlaylist)) {
-        this._renderHLSExoPlayer(url);
-      }
+      hevcRegexp = /CODECS=".*?((hev1)|(hvc1))\..*?"/;
+      masterPlaylist = await (await masterPlaylistPromise).text();
+    }
+    // @ts-ignore
+    if (this._useExoPlayer && hevcRegexp.test(masterPlaylist)) {
```
Comment (bramkragten) [inline src/components/ha-hls-player.ts:135]:
```suggestion
    if (this._useExoPlayer && hevcRegexp!.test(masterPlaylist!)) {
```

**Review by bramkragten — APPROVED** [review]
